File: Source/OldCode/Archive/2004 TVA Code Library/Source/EE/Phasor/PhasorDefinitionBase.vb
'***********************************************************************
'  PhasorDefinitionBase.vb - Phasor value definition base class
'  Copyright © 2005 - TVA, all rights reserved
'
'  Build Environment: VB.NET, Visual Studio 2003
'  Primary Developer: James R Carroll, System Analyst [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  ---------------------------------------------------------------------
'  02/18/2005 - James R Carroll
'       Initial version of source generated
'
'***********************************************************************

Namespace EE.Phasor

    ' This class represents the common implementation of the protocol independent definition of a phasor value.
    Public MustInherit Class PhasorDefinitionBase

        Inherits ChannelDefinitionBase
        Implements IPhasorDefinition

        Private m_type As PhasorType
        Private m_voltageReference As IPhasorDefinition

        Protected Sub New()

            MyBase.New()

            m_type = PhasorType.Voltage
            m_voltageReference = Me

        End Sub

        Protected Sub New(ByVal index As Integer, ByVal label As String, ByVal scale As Integer, ByVal offset As Double, ByVal type As PhasorType, ByVal voltageReference As IPhasorDefinition)

            MyBase.New(index, label, scale, offset)

            m_type = type

            If type = PhasorType.Voltage Then
                m_voltageReference = Me
            Else
                m_voltageReference = voltageReference
            End If

        End Sub

        ' Dervied classes are expected to expose a Public Sub New(ByVal phasorDefinition As IPhasorDefinition)
        Protected Sub New(ByVal phasorDefinition As IPhasorDefinition)

            Me.New(phasorDefinition.Index, phasorDefinition.Label, phasorDefinition.ScalingFactor, phasorDefinition.Offset, _
                phasorDefinition.Type, phasorDefinition.VoltageReference)

        End Sub

        Public Overridable Property [Type]() As PhasorType Implements IPhasorDefinition.Type
            Get
                Return m_type
            End Get
            Set(ByVal Value As PhasorType)
                m_type = Value
            End Set
        End Property

        Public Overridable Property VoltageReference() As IPhasorDefinition Implements IPhasorDefinition.VoltageReference
            Get
                Return m_voltageReference
            End Get
            Set(ByVal Value As IPhasorDefinition)
                If m_type = PhasorType.Voltage Then
                    If Not Value Is Me Then
                        Throw New NotImplementedException("Voltage phasors do not have a voltage reference")
                    End If
                Else
                    m_voltageReference = Value
                End If
            End Set
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/PhasorDefinitionBase.cs
//*******************************************************************************************************
//  PhasorDefinitionBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  02/18/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using System.Security.Permissions;

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent definition of a <see cref="IPhasorValue"/>.
    /// </summary>
    [Serializable()]
    public abstract class PhasorDefinitionBase : ChannelDefinitionBase, IPhasorDefinition
    {
        #region [ Members ]

        // Fields
        private PhasorType m_type;
        private IPhasorDefinition m_voltageReference;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="PhasorDefinitionBase"/>.</param>
        protected PhasorDefinitionBase(IConfigurationCell parent)
            : base(parent)
        {
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinitionBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="label">The label of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="scale">The integer scaling value of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="offset">The offset of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="type">The <see cref="PhasorType"/> of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="voltageReference">The associated <see cref="IPhasorDefinition"/> that represents the voltage reference (if any).</param>
        protected PhasorDefinitionBase(IConfigurationCell parent, string label, uint scale, double offset, PhasorType type, IPhasorDefinition voltageReference)
            : base(parent, label, scale, offset)
        {
            m_type = type;

            if (type == PhasorType.Voltage)
                m_voltageReference = this;
            else
                m_voltageReference = voltageReference;
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinitionBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected PhasorDefinitionBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize phasor definition
            m_type = (PhasorType)info.GetValue("type", typeof(PhasorType));
            m_voltageReference = (IPhasorDefinition)info.GetValue("voltageReference", typeof(IPhasorDefinition));
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="PhasorProtocols.DataFormat"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public override DataFormat DataFormat
        {
            get
            {
                return Parent.PhasorDataFormat;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="PhasorProtocols.CoordinateFormat"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public virtual CoordinateFormat CoordinateFormat
        {
            get
            {
                return Parent.PhasorCoordinateFormat;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="PhasorType"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public virtual PhasorType Type
        {
            get
            {
                return m_type;
            }
            set
            {
                m_type = value;
            }
        }

        /// <summary>
        /// Gets or sets the associated <see cref="IPhasorDefinition"/> that represents the voltage reference (if any).
        /// </summary>
        /// <remarks>
        /// This only applies to current phasors.
        /// </remarks>
        public virtual IPhasorDefinition VoltageReference
        {
            get
            {
                return m_voltageReference;
            }
            set
            {
                if (m_type == PhasorType.Voltage)
                    m_voltageReference = this;
                else
                    m_voltageReference = value;
            }
        }

        /// <summary>
        /// Gets a <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for this <see cref="PhasorDefinitionBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Phasor Type", (int)Type + ": " + Type);

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        [SecurityPermission(SecurityAction.LinkDemand, Flags = SecurityPermissionFlag.SerializationFormatter)]
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize phasor definition
            info.AddValue("type", m_type, typeof(PhasorType));
            info.AddValue("voltageReference", m_voltageReference, typeof(IPhasorDefinition));
        }

        #endregion
    }
}